// @lookai/shared/src/types.ts
import { z } from "zod";

export const MessageRoleSchema = z.enum(["system", "user", "assistant", "tool"]);
export type MessageRole = z.infer<typeof MessageRoleSchema>;

export const ToolCallSchema = z.object({
  id: z.string(),
  name: z.string(),
  arguments: z.record(z.any()),
});
export type ToolCall = z.infer<typeof ToolCallSchema>;

export const MessageSchema = z.object({
  role: MessageRole,
  content: z.string().optional(),
  toolCalls: z.array(ToolCallSchema).optional(),
  toolCallId: z.string().optional(),
  toolName: z.string().optional(),
});
export type Message = z.infer<typeof MessageSchema>;

export const UsageSchema = z.object({
  promptTokens: z.number().int().min(0),
  completionTokens: z.number().int().min(0),
  totalTokens: z.number().int().min(0),
});
export type Usage = z.infer<typeof UsageSchema>;

export const StopReasonSchema = z.enum(["tool_use", "end_turn", "max_tokens", "error"]);
export type StopReason = z.infer<typeof StopReasonSchema>;

export const LLMResponseSchema = z.object({
  stopReason: StopReasonSchema,
  text: z.string(),
  toolCalls: z.array(ToolCallSchema),
  usage: UsageSchema,
});
export type LLMResponse = z.infer<typeof LLMResponseSchema>;

export const LLMOptionsSchema = z.object({
  temperature: z.number().min(0).max(2).optional(),
  maxTokens: z.number().int().positive().optional(),
  model: z.string().optional(),
  responseFormat: z.object({ type: z.literal("json_object") }).optional(),
});
export type LLMOptions = z.infer<typeof LLMOptionsSchema>;

export interface LLMClient {
  create(_messages: Message[], _tools: ToolDef[], _opts?: LLMOptions): Promise<LLMResponse>;
}

export const ToolDefSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.custom<z.ZodTypeAny>((val) => val instanceof z.ZodType),
});
export type ToolDef = z.infer<typeof ToolDefSchema>;

export interface ToolResult {
  ok: boolean;
  content?: string;
  error?: string;
}

export interface Tool {
  name: string;
  description: string;
  parameters: z.ZodTypeAny;
  execute(_args: Record<string, unknown>): Promise<ToolResult> | ToolResult;
}
